### PR TITLE
Added new conditions for the trainer slide-in system

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
Addresses something brought up in #1655; missing conditionals that are present in the official games.

mGBA_20230302_061420748

The new conditions on which a trainer NPC is now allowed to say something are as follows:
-When their last Pokémon's HP is between 25% and 50%.
-When a critical hit is dealt to their first Pokémon for the first time.
-When a super effective hit is dealt to their first Pokémon for the first time.
-When a STAB move is used against (but doesn't faint) the opponent's Pokémon.
-When the Player's Pokémon's move doesn't affect the opponent's Pokémon.

Additionally, I added 2 conditions that are not present in the official games *(as far as I remember anyway.)*
-When they're about to trigger a Mega Evolution.
-When they're about to trigger a Z-Move.
These 2 are extrapolated from the fact that Dynamax/Gigantamax does have a slide-in message in the game where that gimmick is used.
If that one has slide-in messages, why should Megas or Z-Moves not have one too?

## **Discord contact info**
Lunos#4026

### CODE REVIEW

**Comment by Jaizu** [comment]
Please fix conflicts :D

**Comment by LOuroboros** [comment]
> Please fix conflicts :D

Done. I also fixed the order of the `various`es. For whatever reason, `VARIOUS_JUMP_IF_SHELL_TRAP` wasn't followed the order in which the variouses were defined at `include/constants/battle_script_commands.h`.

**Review by DizzyEggg — CHANGES_REQUESTED** [review]
Please fix conflicts.

**Comment by LOuroboros** [comment]
> Please fix conflicts.

Done. Needs testing due to the extra changes from #2699 that got in the way of the changes I made in this PR.

If no one tests things, I'll do it myself once I wake up later.

**Comment by LOuroboros** [comment]
> Checking some Scarlet/Violet videos I saw that they have a trainer message just as they send their first Pokémon. Do you plan to add those too?

I'll look into it.

**Comment by LOuroboros** [comment]
Done.

**Comment by LOuroboros** [comment]
I recorded a .gif because why not, which I forgot to drop because I'm dumb.

mGBA_20230302_061420748

File: src/battle_message.c [hunk 1]
```diff
@@ -3918,14 +3934,6 @@ static u32 GetEnemyMonCount(u32 firstId, u32 lastId, bool32 onlyAlive)
...
-static bool32 IsBattlerHpLow(u32 battler)
-{
-    if ((gBattleMons[battler].hp * 100) / gBattleMons[battler].maxHP < 25)
-        return TRUE;
-    else
-        return FALSE;
-}
-
 u32 ShouldDoTrainerSlide(u32 battlerId, u32 which)
 {
```
Comment (ghoulslash) [inline src/battle_message.c:3938]: can we clean up some of the code below by calling frequently-used functions e.g. `GetEnemyMonCount(firstId, lastId, TRUE)` , or initializing variables like currHp, maxHp for the hp threshold checks?

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -11176,6 +11178,39 @@ static void Cmd_various(void)
         gBattlescriptCurrInstr = cmd->nextInstr;
         return;
     }
+    case VARIOUS_JUMP_IF_SHELL_TRAP:
```
Comment (ghoulslash) [inline src/battle_script_commands.c:11181]: why is this included here?

**Review by ghoulslash — CHANGES_REQUESTED** [review]

File: include/battle.h [hunk 3]
```diff
@@ -622,7 +622,7 @@ struct BattleStruct
     struct MegaEvolutionData mega;
     struct ZMoveData zmove;
     const u8 *trainerSlideMsg;
-    bool8 trainerSlideLowHpMsgDone;
+    bool8 trainerSlideLowHpMsgDone:1;
```
[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/battle.h:625]: Why is that, and what do you suggest me to do? :eyes:

File: include/battle.h [hunk 4]
```diff
@@ -663,6 +663,14 @@ struct BattleStruct
     u8 storedHealingWish:4; // Each battler as a bit.
     u8 storedLunarDance:4; // Each battler as a bit.
     u16 supremeOverlordModifier[MAX_BATTLERS_COUNT];
+    bool8 trainerSlideHalfHpMsgDone:1;
```
[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/battle.h:666]: Why? It's only used as a bool.

File: include/battle.h [hunk 5]
```diff
@@ -663,6 +663,14 @@ struct BattleStruct
...
+    u8 trainerSlideFirstCriticalHitMsgState:2;
+    u8 trainerSlideFirstSuperEffectiveHitMsgState:2;
+    u8 trainerSlideFirstSTABMoveMsgState:2;
+    u8 trainerSlidePlayerMonUnaffectedMsgState:2;
+    bool8 trainerSlideMegaEvolutionMsgDone:1;
```
[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/battle.h:671]: Why though.

File: include/battle.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ghoulslash) [reply include/battle.h:625]: because of struct alignment. you can't align the next field in the middle of a byte, so it will fill in the remaining 7 bits with padding. you can move the field to another bit-field area. If you don't care about saying the space then its fine as is. I just made this comment in the event you were trying to save the 7 bits

File: include/battle.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (ghoulslash) [reply include/battle.h:666]: to be explicit that its sharing the byte with the fields below

File: src/battle_message.c (hunk 1 above)
Comment (LOuroboros) [reply src/battle_message.c:3938]: `GetEnemyMonCount` is already called.

As for the thresholds, I suppose I could, yeah. I'll see what I can come up with.

File: include/battle.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/battle.h:625]: I was. I rather remove the bit allocation and keep the variable a `bool8` than to use the u8 data type for what's a bool variable honestly. I'll do that, if you don't mind.

File: include/battle.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (LOuroboros) [reply include/battle.h:666]: Is it though? The 4 variables below conform a byte on their own, don't they?

File: src/battle_message.c (hunk 1 above)
Comment (LOuroboros) [reply src/battle_message.c:3938]: Check up my ping in the Discord server later. I need your opinion on a function that I wrote in an attempt to make it those maxHp checks more readable.

File: src/battle_script_commands.c (hunk 2 above)
Comment (LOuroboros) [reply src/battle_script_commands.c:11181]: Because I moved it from its original position.

It was written above `VARIOUS_ACTIVATE_TERRAIN_CHANGE_ABILITIES`, even though it's defined below `VARIOUS_HIT_SWITCH_TARGET_FAILED`, so I moved it to keep consistency.

**Comment by LOuroboros** [comment]
Ready for re-review :eyes:

**Review by ghoulslash — APPROVED** [review]
